libraries/AP_BattMonitor/AP_BattMonitor_AD7091R5.cpp

```diff
@@ -0,0 +1,266 @@
+    // @DisplayName: Amps per volt
+    // @Description: Number of amps that a 1V reading on the current sensor corresponds to.
+    // @Units: A/V
+    // @User: Standard
+    AP_GROUPINFO("AMP_PERVLT", 33, AP_BattMonitor_AD7091R5, _curr_amp_per_volt, 0),
+
+    // @Param: AMP_OFFSET
+    // @DisplayName: AMP offset
+    // @Description: Voltage offset at zero current on current sensor
+    // @Units: V
+    // @User: Standard
+    AP_GROUPINFO("AMP_OFFSET", 34, AP_BattMonitor_AD7091R5, _curr_amp_offset, 0),
+
+    // @Param: VLT_OFFSET
+    // @DisplayName: Volage offset
+    // @Description: Voltage offset on voltage pin. This allows for an offset due to a diode. This voltage is subtracted before the scaling is applied
+    // @Units: V
+    // @User: Advanced
+    AP_GROUPINFO("VLT_OFFSET", 35, AP_BattMonitor_AD7091R5, _volt_offset, 0),
+
+    // Param indexes must be less than 10 to avoid conflict with other battery monitor param tables loaded by pointer
+
+    AP_GROUPEND
+};
+
+
+//Variable initialised to read from first instance.
+AP_BattMonitor_AD7091R5::AnalogData AP_BattMonitor_AD7091R5::_analog_data[AD7091R5_NO_OF_CHANNELS] = {0};
+bool AP_BattMonitor_AD7091R5::_first = true;
+bool AP_BattMonitor_AD7091R5::_health = false;
+
+/**
+ * @brief Construct a new ap battmonitor ad7091r5::ap battmonitor ad7091r5 object
+ *
+ * @param mon
+ * @param mon_state
+ * @param params
+ */
+AP_BattMonitor_AD7091R5::AP_BattMonitor_AD7091R5(AP_BattMonitor &mon,
+        AP_BattMonitor::BattMonitor_State &mon_state,
+        AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _state.var_info = var_info;
+}
+
+/**
+ * @brief probe and initialize the sensor and register call back
+ *
+ */
+void AP_BattMonitor_AD7091R5::init()
+{
+
+    _state.healthy = false;
+    // voltage and current pins from params and check if there are in range
+    if (_volt_pin.get() >= AD7091R5_BASE_PIN && _volt_pin.get() <= AD7091R5_BASE_PIN + AD7091R5_NO_OF_CHANNELS) {
+        volt_buff_pt = _volt_pin.get() - AD7091R5_BASE_PIN;
+
+        if (_curr_pin.get() >= AD7091R5_BASE_PIN && _curr_pin.get() <= AD7091R5_BASE_PIN + AD7091R5_NO_OF_CHANNELS) {
+            curr_buff_pt = _curr_pin.get() - AD7091R5_BASE_PIN;
+            _state.healthy = true; //set status healthy
+        }
+    }
+
+    //return if configured incorrectly
+    if (!_state.healthy) {
+        return;
+    }
+
+    // only the first instance read the i2c device
+    if (_first) {
+        _first = false;
+        _state.healthy = false;
+        // probe i2c device
+        _dev = hal.i2c_mgr->get_device(AD7091R5_I2C_BUS, AD7091R5_I2C_ADDR);
+
+        if (_dev) {
+            WITH_SEMAPHORE(_dev->get_semaphore());
+            _dev->set_retries(10); // lots of retries during probe
+            //Reset and config device
+            if (_reset() && _config()) {
+                _state.healthy = true;
```

Comment:
Don't set healthy here. Healthy implies we filled in the state. That should only be done by read(), as there won't be any valid data until after we've read it. You also don't have to set it false, because that is already handled elsewhere.